dask/multiprocessing.py

```diff
@@ -4,9 +4,27 @@
 from .optimize import fuse, cull
 import multiprocessing
 import dill
+import cloudpickle
 import pickle
 from .async import get_async # TODO: get better get
 from .context import _globals
+from sys import getrecursionlimit, setrecursionlimit
+
+def _dumps(x):
+    try:
+        return cloudpickle.dumps(x, protocol=pickle.HIGHEST_PROTOCOL)
+    except:
+        try:
+            recursionlimit = getrecursionlimit()
+            setrecursionlimit(2*recursionlimit)
+            dumped = cloudpickle.dumps(x, protocol=pickle.HIGHEST_PROTOCOL)
+            setrecursionlimit(recursionlimit)
+            return dumped
+        except:
+            raise
```

Comment:
I have narrowed this down a little bit.   For example in `dask/bag/tests/test_bag.py::test_fold FAILED` the object that is failing to pickle is 

`['bag-33', (<built-in function reduce>, <method 'union' of 'set' objects>, [('bag-32', 0), ('bag-32', 1), ('bag-32', 2)]), {('bag-32', 0): set([0, 1]), ('bag-32', 1): set([2, 3]), ('bag-32', 2): set([4])}, <AutoProxy[Queue] object, typeid 'Queue' at 0x4497110>, <function _process_get_id at 0x34eb410>, False]`

I went ahead and tried pickling the items in this list, and found that the one that causes the recursion error is:

`(<built-in function reduce>, <method 'union' of 'set' objects>, [('bag-32', 0), ('bag-32', 1), ('bag-32', 2)])`

I did the same for the other failing tests, and found that the objects that hit the recursion limit are: 

* `dask/bag/tests/test_bag.py::test_fold FAILED`:
`(<built-in function reduce>, <method 'union' of 'set' objects>, [('bag-32', 0), ('bag-32', 1), ('bag-32', 2)])`

* `dask/bag/tests/test_bag.py::test_distinct FAILED`:
`(<built-in function apply>, <method 'union' of 'set' objects>, [('bag-34', 2), ('bag-34', 1), ('bag-34', 0)])`

* `dask/bag/tests/test_bag.py::test_from_filenames_large FAILED`:
`(<method 'decode' of 'str' objects>, (<function textblock at 0x2157e60>, '/tmp/tmpju03wI.', 200, 300, None), 'utf-8')`

* `dask/bag/tests/test_bag.py::test_from_filenames_encoding FAILED`:
`(<method 'decode' of 'str' objects>, (<function textblock at 0x2157e60>, '/tmp/tmpTsLX9O.', 0, 100, None), 'gb18030')`

* `dask/bag/tests/test_bag.py::test_from_filenames_large_gzip FAILED`:
`(<method 'decode' of 'str' objects>, (<function textblock at 0x2157e60>, '/tmp/tmpKXpz29.gz', 500, 600, 'gzip'), 'utf-8')`